presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ApplyConnectorOptimization.java

```diff
@@ -0,0 +1,266 @@
+        requireNonNull(symbolAllocator, "symbolAllocator is null");
+        requireNonNull(idAllocator, "idAllocator is null");
+
+        if (connectorOptimizers.isEmpty()) {
+            return plan;
+        }
+
+        ConnectorSession connectorSession = session.toConnectorSession();
+
+        // retrieve all the connectors
+        ImmutableSet.Builder<ConnectorId> builder = ImmutableSet.builder();
+        getAllConnectorIds(plan, builder);
+
+        // only retain those having optimizers
+        Set<ConnectorId> connectorIds = new HashSet<>(builder.build());
+        connectorIds.retainAll(connectorOptimizers.keySet());
+        connectorIds = ImmutableSet.copyOf(connectorIds);
+        if (connectorIds.isEmpty()) {
+            // there is no connector optimizer applicable to the current plan
+            return plan;
+        }
+
+        // for each connector, retrieve the set of subplans to optimize
+        // TODO: what if a new connector is added by an existing one
+        for (ConnectorId connectorId : connectorIds) {
+            Set<ConnectorPlanOptimizer> optimizers = connectorOptimizers.get(connectorId);
+
+            Map<PlanNode, ConnectorPlanNodeContext> contextMap = new HashMap<>();
+            buildConnectorPlanNodeContext(plan, null, contextMap);
+
+            // keep track a set of changed nodes; the keys are original nodes and the values are the new nodes
+            Map<PlanNode, PlanNode> updates = new HashMap<>();
+
+            // process connector optimizers
+            for (PlanNode node : contextMap.keySet()) {
+                // get max closure
+                ConnectorPlanNodeContext context = contextMap.get(node);
+                if (!context.isClosure(connectorId) ||
+                        !context.getParent().isPresent() ||
+                        contextMap.get(context.getParent().get()).isClosure(connectorId)) {
+                    continue;
+                }
+
+                PlanNode newNode = node;
+
+                // TODO: the returned node is still a max closure only if there is no new connector added
+                for (ConnectorPlanOptimizer optimizer : optimizers) {
+                    newNode = optimizer.optimize(newNode, connectorSession, symbolAllocator, idAllocator);
+                }
+
+                if (node != newNode) {
+                    // the optimizer has allocated a new PlanNode
+                    checkState(
+                            containsAll(ImmutableSet.copyOf(newNode.getOutputVariables()), node.getOutputVariables()),
+                            format("the connector optimizer from %s returns a node that does not cover all output before optimization", connectorId));
+                    updates.put(node, newNode);
+                }
+            }
+            // up to this point, we have a set of updated nodes; need to recursively update their parents
+
+            // alter the plan with a bottom-up approach
```

Comment:
Update: It looks to me bottom-up is a nice to have but not required (might be less efficient since the same node will be traversed multiple times). See https://github.com/prestodb/presto/pull/12920#discussion_r294464759 .

In this case, maybe update the comment to reflect that `bottom-up` is nice to have, but not required to make algorithm correct?